zanata-war/src/main/java/org/zanata/service/impl/SlugEntityUpdatedListener.java

```diff
@@ -0,0 +1,115 @@
+package org.zanata.service.impl;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.hibernate.event.spi.PostUpdateEvent;
+import org.hibernate.event.spi.PostUpdateEventListener;
+import org.jboss.seam.ScopeType;
+import org.jboss.seam.annotations.AutoCreate;
+import org.jboss.seam.annotations.In;
+import org.jboss.seam.annotations.Name;
+import org.jboss.seam.annotations.Scope;
+import org.zanata.async.AsyncTaskHandle;
+import org.zanata.async.AsyncTaskHandleManager;
+import org.zanata.events.ProjectIterationUpdate;
+import org.zanata.events.ProjectUpdate;
+import org.zanata.model.HProject;
+import org.zanata.model.HProjectIteration;
+import org.zanata.model.SlugEntityBase;
+import org.zanata.service.IndexingService;
+import org.zanata.util.Event;
+
+/**
+ * This class is a hibernate event listener which listens on post commit events
+ * for HProject and HProjectIteration. If it detects a HProject slug change, it
+ * will perform re-indexing for all HTextFlowTargets under that project. It will
+ * also fire update event for HProject and HProjectIteration with their old slug
+ * in payload.
+ *
+ * @see org.zanata.webtrans.server.HibernateIntegrator
+ * @see org.zanata.webtrans.server.TranslationWorkspaceManagerImpl
+ * @see IndexingServiceImpl
+ *
+ * @author Patrick Huang <a
+ *         href="mailto:pahuang@redhat.com">pahuang@redhat.com</a>
+ */
+@Name("slugEntityUpdatedListener")
+@Scope(ScopeType.APPLICATION)
```

Comment:
The index itself is not a problem. But it does reminded me a potential bug. It used to only handle HProject but then it changed to also handle HProjectIteration. it needs two indexes instead of one. I will fix it.